wagtail/wagtail
# 🎛️ Replace base `ExpandingFormset` functionality with Stimulus FormsetController `w-formset` inc. initial adoption

First stage of #7915

## Overview

* Introduces a new Stimulus class `FormsetController` with the identifier `w-formset` with the longer term plan for this to be used for the InlinePanel JS functionality also.
* Introduces a refactor of our common code for Django's formsets that streamlines how we set the delete input widget and prepares some of the Stimulus data attributes for use in the templates. This is a new base mixin `wagtail.admin.forms.formsets.BaseFormSetMixin` (happy for a different name to be suggested).
* Replaces only the basic expanding formset usage in Group edit/create views, Workflow edit/create (Page selection only) with the Stimulus usage. No changes to `InlinePanel` at all.
* Most importantly, this new approach is CSP compliant, has validation error messaging feedback for developers and does not rely on any complex id string wrangling, with fully tested support for complex nested usage.
  * Stimulus does most of this for us through the [Stimulus Scopes](https://stimulus.hotwired.dev/reference/controllers#scopes) feature.
  * I've left all nested IDs in place in the templates as there is no major need to remove these for now, but they're no longer used for any JS behaviour.
  * Relates to getting closer to CSP compliance #7053 & #1288

### Additional notes

* Introduces event dispatching for the expanding formsets to align with the `InlinePanel` https://docs.wagtail.org/en/stable/reference/pages/panels.html#inline-panel-events (have not documented, I thought release notes would be enough for this as it's not really used by customised code).
* Removes the reliance on jQuery animations by introducing a function (which we can abstract later if it's useful) that uses a promise callback for 'hiding' after an animation/transition is completed on an element without having to hard-code how long we expect the animation to take. For example, we can change the `selected` class to be ...-200 and the hidden attribute will be set after 200ms instead of the current 300. This reduces a common issue where we get the animation timing and the attribute updating out of sync causing chunky animations or confusing delays.
* Commits are broken up to be a step by step creation and adoption of this new code, hopefully making reviewing easier.
* instead of style display none, now using `hidden` for when a child form is deleted. This is more appropriate for the UI as the element is essentially meant to be removed. This also makes CSP issues with style attribute changes go away.
* There will be a subtle change in behaviour between the current, if you add 1001 children in any of the usage below. Currently, this would fail on server submit, however, with this change the add button will simply not add any more. I can change this to introduce a new option to ignore min/max but it seemed to make sense to keep this initial adoption simpler even with that edge case behaviour difference.
* Naming of 'formset' vs 'form_set'
  * I have opted for `FormSet` in the Python class names, this is what Django uses.
  * `formsets.py` for the file in our admin files, again, following the Django convention.
  * `FormsetController`/`w-formset` (not `FormSetController`/`w-form-set`) in the Javascript code for brevity and alignment with our existing set up event names on the InlinePanel code (e.g. `w-formset:added`).
* Naming of the `'removed'` event vs (`delete`/`deleted` action/code).
  * Django uses the convention `removed` for the event https://docs.djangoproject.com/en/5.0/ref/contrib/admin/javascript/ we use that same event name in our documentation for InlinePanel https://docs.wagtail.org/en/stable/reference/pages/panels.html#javascript-dom-events
  * Django uses the names `delete`/`deleted` in the code.
  * I have aligned with that approach in the formset, while it's a bit odd to have `removed`/`removing` in the event names I hope this clarifies why.
* Additional clean up benefits include reduction of duplicate code that was imported across views, better leverage of Python / Django mixins to support potential Python side customisation behaviour in the future, and increased Python test coverage over the data attribute generation.

## Checklist

-   [X] For front-end changes: Did you test on all of Wagtail’s supported environments? Chrome, Safari, Firefox all manually tested
-   [X] For new features: Has the documentation been updated accordingly? (Upgrade considerations added)

## How to test

1. Workflow edit, workflow create - validate that the pages sections all update correctly when adding/removing.
2. Workflow edit for a disabled workflow, check the console has no errors and the content is not visible for page selection as per current code.
3. User group create/edit - validate the page permissions and other non-page permissions changes.
4. For the above, check things still work as expected on Save with a POST form failure.


## Proposed upgrade considerations

### Deprecation of `window.buildExpandingFormset` global function

The global function `window.buildExpandingFormset` to attach basic JavaScript behavior for Django's expanding formsets has been deprecated and will be removed in a future release.

Using `InlinePanel` will not be impacted, this only impacts a small set of basic expanding formsets in use across Workflow and Group view editing.

Previously these expanding formsets required a mix of specific id attribute structures and inline scripts to instantiate with callbacks for handling deletion.

In the future, the following data attributes are required to emulate the same behavior, these are likely to change and should not be considered official documentation.

| Element                                      | Attribute(s)                              |
| -------------------------------------------- | ----------------------------------------- |
| Containing element                           | `data-controller="w-formset"`             |
| Element to append new child forms            | `data-w-formset-target="forms"`           |
| Child form element                           | `data-w-formset-target="child"`           |
| Deleted form element                         | `data-w-formset-target="deleted" hidden`  |
| `template` element for blank form            | `data-w-formset-target="template"`        |
| Management field (total forms)               | `data-w-formset-target="totalFormsInput"` |
| Management field (min forms)                 | `data-w-formset-target="minFormsInput"`   |
| Management field (max forms)                 | `data-w-formset-target="maxFormsInput"`   |
| Management field (Delete, within child form) | `data-w-formset-target="deleteInput"`     |
| Add child `button`                           | `data-action="w-formset#add"`             |
| Delete child `button` (within child form)    | `data-action="w-formset#delete"`          |

Usage of nested id structures are no longer required but can be left in place for easier debugging.
